File: src/Go/burn_test/burn_tester.go
package burn_test

import (
	. "../common"
	. "../common/nanotime"
	"fmt"
	"math"
	"runtime"
	"strconv"
	"time"
)

const ArrayItemSize = 8
const ArrayItemSizeLog2 = 3
const MinArraySize = 24
const MinAllocationSize = MinArraySize + 8

const AllocationSequenceLength = 1 << 20 // 2^20, i.e. ~ 1M items; must be a power of 2!
const MinSize = 1
const MaxSize = 1 << 17 // 128KB
const TimeSamplerFrequency = 1000000
const MaxTime = 1000 * TimeSamplerFrequency // 1000 seconds

var MinGCPause = time.Duration(10 * time.Microsecond).Nanoseconds()
var DefaultDuration = time.Duration(10 * time.Second)

type BurnTester struct {
	NoOutput       bool
	Duration       time.Duration
	StaticSetSize  int64
	Allocations    []AllocationInfo
	StartIndexes   []int32
	StaticSet      [][][]int64
	StaticSetCount int64
	Random         *StdRandom
	isInitialized  bool
}

func NewBurnTester(staticSetSize int64) *BurnTester {
	t := &BurnTester{
		NoOutput:       false,
		Duration:       DefaultDuration,
		StaticSetSize:  staticSetSize,
		Allocations:    make([]AllocationInfo, AllocationSequenceLength),
		StartIndexes:   make([]int32, AllocationSequenceLength),
		StaticSet:      make([][][]int64, 0),
		StaticSetCount: 0,
		Random:         NewStdRandom(123),
	}
	return t
}

func NewWarmupBurnTester(staticSetSize int64) *BurnTester {
	t := NewBurnTester(staticSetSize)
	t.NoOutput = true
	t.Duration = time.Duration(1 * time.Second)
	return t
}

func (t *BurnTester) TryInitialize() {
	if t.isInitialized {
		return
	}

	sizeSampler := Truncate(CreateStandardSizeSampler(t.Random), MinSize, MaxSize)
	timeSampler := Truncate(CreateStandardTimeSampler(t.Random), 0, MaxTime)
	releaseCycleTimeInSeconds := NanosecondsPerReleaseCycle * Nano
	timeFactor := (1.0 / TimeSamplerFrequency) / releaseCycleTimeInSeconds
	for i := 0; i < AllocationSequenceLength; i++ {
		size := int32(sizeSampler.Sample())
		arraySize := ByteSizeToArraySize(size)

		_time := timeSampler.Sample()
		_timeStr := strconv.Itoa(int(_time * timeFactor))
		bucketIndex := int8(len(_timeStr) - 1)
		generationIndex := int8(_timeStr[0] - '0')

		t.Allocations[i] = AllocationInfo{arraySize, bucketIndex, generationIndex}
	}

	for i := 0; i < ThreadCount; i++ {
		t.StartIndexes[i] = int32(t.Random.Next()) % AllocationSequenceLength
	}

	startOffset := int32(t.Random.Next()) % AllocationSequenceLength
	runner := NewParallelRunner(func(i int) IActivity {
		return NewSetAllocator(
			t.StaticSetSize/int64(ThreadCount),
			t.Allocations,
			(startOffset+t.StartIndexes[i])%AllocationSequenceLength)
	})
	activities := runner.Run()
	for _, a := range activities {
		allocator := a.(*SetAllocator)
		t.StaticSet = append(t.StaticSet, allocator.Set)
		t.StaticSetCount += int64(len(allocator.Set))
	}

	runtime.GC()

	t.isInitialized = true
}

func (t *BurnTester) Run() {
	t.TryInitialize()

	duration := t.Duration.Seconds()
	if !t.NoOutput {
		fmt.Printf("Test settings:\n")
		fmt.Printf("  Duration:          %v s\n", int(duration))
		fmt.Printf("  Thread count:      %v\n", ThreadCount)
		fmt.Printf("  Static set:\n")
		fmt.Printf("    Total size:      %.3f GB\n", float64(t.StaticSetSize)/GB)
		fmt.Printf("    Object count:    %.3f M\n", float64(t.StaticSetCount)/Mega)
		fmt.Println()
	}

	runner := NewParallelRunner(func(i int) IActivity { return NewGarbageAllocator(t.Duration, t.Allocations, t.StartIndexes[i]) })

	msPre := new(runtime.MemStats)
	runtime.ReadMemStats(msPre)
	startTime := float64(Nanotime().Nanoseconds())
	activities := runner.Run()
	msPost := new(runtime.MemStats)
	runtime.ReadMemStats(msPost)

	// Slice item casting
	allocators := make([]*GarbageAllocator, 0)
	for _, a := range activities {
		allocators = append(allocators, a.(*GarbageAllocator))
	}

	if t.NoOutput {
		return
	}

	const NanoToMilliseconds = 1000 / Giga // ns to ms

	// Normalizing GCPauses; nanoseconds to milliseconds and a bit more
	var pauses [][]Interval
	for _, a := range allocators {
		for j := range a.GCPauses {
			a.GCPauses[j].Start = (a.GCPauses[j].Start - startTime) * NanoToMilliseconds
			a.GCPauses[j].End = (a.GCPauses[j].End - startTime - 0.5) * NanoToMilliseconds

		}
		pauses = append(pauses, ToCanonicalSorted(a.GCPauses))
	}
	var intersections = pauses[0]
	for _, p := range pauses {
		intersections = IntersectSortedPairs(intersections, p)
	}
	var globalPauses []float64
	var globalPausesSum float64
	for _, p := range intersections {
		globalPauses = append(globalPauses, p.End-p.Start)
		globalPausesSum += p.End - p.Start
	}
	var allocationSizes []float64
	var allocationHoldDurations []float64
	for _, a := range allocators {
		for _, ai := range a.Allocations {
			t := math.Pow(10, float64(ai.BucketIndex)) * float64(ai.GenerationIndex) * NanosecondsPerReleaseCycle * NanoToMilliseconds
			allocationHoldDurations = append(allocationHoldDurations, t)
			allocationSizes = append(allocationSizes, float64(ArraySizeToByteSize(ai.ArraySize)))
		}
	}

	fmt.Printf("Allocation speed:\n")
	var ops, bytes int64
	for _, a := range allocators {
		ops += a.AllocationCount
		bytes += a.ByteCount
	}
	fmt.Printf("  Operations per second: %.2f M/s\n", float64(ops)/duration/Mega)
	fmt.Printf("  Bytes per second:      %.2f GB/s\n", float64(bytes)/duration/GB)
	fmt.Printf("  Allocation stats:\n")
	fmt.Printf("    Size:\n")
	DumpArrayStats(allocationSizes, "B", "      ", true)
	fmt.Printf("    Hold duration:\n")
	DumpArrayStats(allocationHoldDurations, "ms", "      ", true)
	fmt.Println()
	fmt.Printf("GC stats:\n")
	fmt.Printf("  RAM used:              %.3f -> %.3f GB\n", float64(msPre.HeapAlloc)/GB, float64(msPost.HeapAlloc)/GB)
	fmt.Printf("  GC rate:               %.3f /s\n", float64(msPost.NumGC-msPre.NumGC)/duration)
	fmt.Printf("  Allocation rate:\n")
	fmt.Printf("    Objects:             %.3f M/s, %.3f M/s freed\n",
		float64(msPost.Mallocs-msPre.Mallocs)/duration/Mega,
		float64(msPost.Frees-msPre.Frees)/duration/Mega)
	fmt.Printf("    Bytes:               %.3f GB/s\n", float64(msPost.TotalAlloc-msPre.TotalAlloc)/duration/GB)
	fmt.Printf("  Global pauses:\n")
	fmt.Printf("    %% of time frozen:    %.3f %%, %.3f %% as reported by runtime\n",
		globalPausesSum/1000/duration*100,
		float64(msPost.PauseTotalNs-msPre.PauseTotalNs)/Giga/duration*100)
	fmt.Printf("    # per second:        %.3f /s\n", float64(len(globalPauses))/duration)
	DumpArrayStats(globalPauses, "ms", "      ", true)
	fmt.Println()
}

File: src/CSharp/Main/BurnTest/BurnTester.cs
using System;
using System.CodeDom.Compiler;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Runtime;
using Benchmarking.Common;

namespace GCBurn.BurnTest 
{
    public class BurnTester
    {
        public const int AllocationSequenceLength = 1 << 20; // 2^20, i.e. ~ 1M items; must be a power of 2!
        public const double TimeSamplerFrequency = 1_000_000;
        public const double TimeSamplerUnitInSeconds = 1 / TimeSamplerFrequency;
        public static TimeSpan DefaultDuration = TimeSpan.FromSeconds(10);
        public static double DefaultMaxTime = 1000 * TimeSamplerFrequency; // 1000 seconds
        public static double DefaultMaxSize = 1 << 17; // 128KB
        
        public TimeSpan Duration = DefaultDuration;
        public StdRandom Random = new StdRandom(123); 
        public Func<StdRandom, IDistribution> CreateSizeSampler = Samplers.CreateStandardSizeSampler; 
        public Func<StdRandom, IDistribution> CreateTimeSampler = Samplers.CreateStandardTimeSampler;
        public double MaxTime = DefaultMaxTime;
        public double MaxSize = DefaultMaxSize;
        public long StaticSetSize = 0;
        public IndentedTextWriter Writer = new IndentedTextWriter(Console.Out, "  ");
        
        private bool _isInitialized;
        private object[][] _staticSet;
        private (int ArraySize, sbyte BucketIndex, sbyte GenerationIndex)[] _allocations;
        private int[] _startIndexes;
        
        public static BurnTester New(long staticSetSize) => new BurnTester() {
            StaticSetSize = staticSetSize
        };
        
        public static BurnTester NewWarmup(long staticSetSize) => new BurnTester() {
            StaticSetSize = staticSetSize,
            Duration = TimeSpan.FromSeconds(1), 
            Writer = new IndentedTextWriter(TextWriter.Null),
        };

        public void TryInitialize()
        {
            if (_isInitialized)
                return;
            
            _allocations = new (int, sbyte, sbyte) [AllocationSequenceLength];
            var sizeSampler = CreateSizeSampler(Random).Truncate(1, MaxSize);
            var timeSampler = CreateTimeSampler(Random).Truncate(0, MaxTime);
            var releaseCycleTimeInSeconds = 1.0 * GarbageHolder.TicksPerReleaseCycle / Stopwatch.Frequency;
            var timeFactor = TimeSamplerUnitInSeconds / releaseCycleTimeInSeconds;
            for (var i = 0; i < AllocationSequenceLength; i++) {
                var size = (int) sizeSampler.Sample();
                var arraySize = GarbageAllocator.ByteSizeToArraySize(size);
                
                var time = timeSampler.Sample();
                var timeStr = ((long) (time * timeFactor)).ToString();
                var bucketIndex = (sbyte) (timeStr.Length - 1);
                var generationIndex = (sbyte) (timeStr[0] - '0');
                
                _allocations[i] = (arraySize, bucketIndex, generationIndex);
            }

            _startIndexes = Enumerable.Range(0, ParallelRunner.ThreadCount)
                .Select(_ => Random.Next() % AllocationSequenceLength)
                .ToArray();
            
            // Creating _staticSet
            var startOffset = Random.Next() % AllocationSequenceLength;
            _staticSet = ParallelRunner.New(i => new SetAllocator(
                    StaticSetSize / ParallelRunner.ThreadCount,
                    _allocations,
                    (_startIndexes[i] + startOffset) % AllocationSequenceLength)
                )
                .Run()
                .SelectMany(a => a.Set)
                .ToArray();
            
            GC.Collect(); // To make it uniform w/ Go test

            _isInitialized = true;
        }

        public void Run()
        {
            TryInitialize();
            
            var testDuration = Duration.TotalSeconds;
            var threadCount = ParallelRunner.ThreadCount;
            using (Writer.Section($"Test settings:")) {
                Writer.AppendMetric("Duration", testDuration, "s");
                Writer.AppendMetric("Thread count", threadCount, "");
                using (Writer.Section($"Static set:")) {
                    Writer.AppendMetric("Total size", StaticSetSize / Sizes.GB, "GB");
                    Writer.AppendMetric("Object count", _staticSet.Sum(a => (long) a.Length) / Sizes.Mega, "M");
                }
            }

            var runner = ParallelRunner.New(i => new GarbageAllocator(Duration, _allocations, _startIndexes[i]));
            var gcCounts = GCInfo.GetGCCounts();
            var memUsedBefore = GC.GetTotalMemory(false);
            var startTimestamp = Stopwatch.GetTimestamp();
            var allocators = runner.Run();
            var memUsedAfter = GC.GetTotalMemory(false);

            Writer.AppendLine();

            // Normalizing GCPauses (converting time to ms, removing offset)
            var ticksToMilliseconds = 1000.0 / Stopwatch.Frequency;
            foreach (var a in allocators) {
                a.GCPauses = a.GCPauses
                    .Select(p => p
                        .MoveBy(-startTimestamp)
                        .ExpandBy(-0.5)
                        .ScaleBy(ticksToMilliseconds))
                    .ToCanonicalSorted()
                    .ToList();
            }

            // Computing thread & global pauses
            var allPauses = allocators.SelectMany(a => a.GCPauses);
            var duration = allPauses.Max(p => p.End) / 1000; // In seconds (as testDuration)
            var threadPauses = allocators
                .Select(a => a.GCPauses
                    .Select(p => p.Size)
                    .OrderBy(d => d)
                    .ToArray())
                .ToArray();
            var pauseCountPerSecondPerThread = allocators
                .SelectMany(a => a.GCPauses
                    .GroupBy(p => Math.Floor(p.Start / 1000))
                    .Select(g => (double) g.Count())
                )
                .OrderBy(c => c)
                .ToArray();
            var globalPauses = allocators
                .Select(a => a.GCPauses)
                .IntersectSortedTuples()
                .Select(p => p.Size)
                .OrderBy(d => d)
                .ToArray();
            var allocationSizes = _allocations
                .Select(a => (double) GarbageAllocator.ArraySizeToByteSize(a.ArraySize))
                .OrderBy(s => s)
                .ToArray();
            var allocationHoldDurations = _allocations
                .Select(a => Math.Pow(10, a.BucketIndex) * a.GenerationIndex 
                    * GarbageHolder.TicksPerReleaseCycle * ticksToMilliseconds)
                .OrderBy(s => s)
                .ToArray();

            Writer.AppendMetric("Actual duration", duration, "s");
            using (Writer.Section($"Allocation speed:")) {
                Writer.AppendMetric("Operations per second", allocators.Sum(a => a.AllocationCount) / duration / Sizes.Mega, "M/s");
                Writer.AppendMetric("Bytes per second", allocators.Sum(a => a.ByteCount)  / duration / Sizes.GB, "GB/s");
                using (Writer.Section($"Allocation stats:")) {
                    Writer.AppendHistogram("Size:", allocationSizes, "B");
                    Writer.AppendHistogram("Hold duration:", allocationHoldDurations, "ms");
                }
            }
            using (Writer.Section($"GC stats:")) {
                Writer.AppendValue("RAM used", $"{(memUsedBefore / Sizes.GB):0.###} -> {(memUsedAfter / Sizes.GB):0.###} GB");
                if (!GCSettings.IsServerGC) {
                    // This data doesn't seem to be correct when ServerGC is on
                    using (Writer.Section($"GC rate:")) {
                        foreach (var (c, i) in gcCounts.WithIndexes())
                            Writer.AppendMetric($"Gen{i}, # per second", c / duration, "/s");
                    }
                }
                using (Writer.Section($"Thread pauses:")) {
                    Writer.AppendMetric("% of time frozen", threadPauses.SelectMany(p => p).Sum() / 1000 / threadCount / duration * 100, "%");
                    Writer.AppendHistogram("# per second:", pauseCountPerSecondPerThread, "/s");
                }
                using (Writer.Section($"Global pauses:")) {
                    Writer.AppendMetric("% of time frozen", globalPauses.Sum() / 1000 / duration * 100, "%");
                    Writer.AppendMetric("# per second", globalPauses.Length / duration, "/s");
                    Writer.AppendHistogram("Pause duration:", globalPauses, "ms");
                }
            }
            Writer.AppendLine();
        }
    }
}
